From f3cd79e048c5beeb3d2720256017d93872f2c94a Mon Sep 17 00:00:00 2001
From: Bjorn Andersson <bjorn.andersson@sonymobile.com>
Date: Tue, 6 May 2014 22:16:16 -0700
Subject: [PATCH] ARM: zImage: Allow DTB to override broken ATAG_MEM

Support overriding ATAG_MEM, by specifying non-zero content of the /memory/reg
property in the appended DTB. This is needed to work around bootloaders passing
broken tags.

Signed-off-by: Bjorn Andersson <bjorn.andersson@sonymobile.com>
---
 arch/arm/boot/compressed/atags_to_fdt.c | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/arch/arm/boot/compressed/atags_to_fdt.c b/arch/arm/boot/compressed/atags_to_fdt.c
index 9448aa0..34b448c 100644
--- a/arch/arm/boot/compressed/atags_to_fdt.c
+++ b/arch/arm/boot/compressed/atags_to_fdt.c
@@ -97,6 +97,22 @@ static void merge_fdt_bootargs(void *fdt, const char *fdt_cmdline)
 	setprop_string(fdt, "/chosen", "bootargs", cmdline);
 }
 
+static int fdt_overrides_atag_mem(void *fdt)
+{
+	const char *memory;
+	int len = 0;
+
+	memory = getprop(fdt, "/memory", "reg", &len);
+	if (memory) {
+		while (len--) {
+			if (*memory != '\0')
+				return 1;
+		}
+	}
+
+	return 0;
+}
+
 /*
  * Convert and fold provided ATAGs into the provided FDT.
  *
@@ -182,7 +198,7 @@ int atags_to_fdt(void *atag_list, void *fdt, int total_space)
 		}
 	}
 
-	if (memcount) {
+	if (memcount && !fdt_overrides_atag_mem(fdt)) {
 		setprop(fdt, "/memory", "reg", mem_reg_property,
 			4 * memcount * memsize);
 	}
-- 
2.0.2

